File: LeetCode/Medium/1615-maximal-network-rank/1615-maximal-network-rank.py
class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        degrees = [0] * n

        for road in roads:
            degrees[road[0]] += 1
            degrees[road[1]] += 1
        roads_set = set(tuple(road) for road in roads)
        max_rank = 0
        for i in range(n):
            for j in range(i+1,n):
                rank = degrees[i] + degrees[j]
                if (i,j) in roads_set or (j,i) in roads_set:
                    rank -= 1
                max_rank = max(max_rank, rank)
        return max_rank

File: LeetCode/Medium/1615-maximal-network-rank/1615-maximal-network-rank.cpp
class Solution {
public:
    int maximalNetworkRank(int n, vector<vector<int>>& roads) {
        vector<int> degrees(n,0);
        vector<vector<bool>> connected(n, vector<bool>(n, false));

        for(int i=0; i<roads.size(); i++){
            degrees[roads[i][0]]++;
            degrees[roads[i][1]]++;
            connected[roads[i][0]][roads[i][1]] = true;
            connected[roads[i][1]][roads[i][0]] = true;
        }
        int max_rank = 0;
        for(int i=0; i<n;i++){
            for(int j=i+1; j<n; j++){
                int rank = degrees[i] + degrees[j];
                if(connected[i][j])
                    rank -= 1;
                max_rank = std::max(max_rank, rank);
            }
        }
        return max_rank;
    }
};